source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs

```diff
@@ -264,6 +265,81 @@ public async Task QueueAndWait_Can_Queue_Dequeue_Apply_VeryLargeNumberOfRequests
             await ApplyResponsesConcurrentlyAndEnsureAllQueueResponsesMatch(sut, requestsInOrder, queueAndWaitTasksInOrder);
         }
 
+        [Test]
+        [NonParallelizable]
+        public async Task QueueAndWait_Can_Queue_Dequeue_WhenRequestsAreBeingCancelled()
+        {
+            // Arrange
+            const string endpoint = "poll://endpoint001";
+
+            var sut = new PendingRequestQueueBuilder().WithSyncOrAsync(SyncOrAsync.Async).WithEndpoint(endpoint).Build();
+            
+            var requestsInOrder = Enumerable.Range(0, 500)
+                .Select(_ => new RequestMessageBuilder(endpoint).Build())
+                .ToList();
+
+            // Act
+            var queueAndWaitTasksInOrder = requestsInOrder
+                .Select(request =>
+                {
+                    var requestCancellationTokenSource = new CancellationTokenSource();
+                    return new Tuple<Task<ResponseMessage>, CancellationTokenSource>(
+                        StartQueueAndWait(sut, request, SyncOrAsync.Async, requestCancellationTokenSource.Token), 
+                        requestCancellationTokenSource);
+                })
+                .ToList();
+
+            await WaitForQueueCountToBecome(sut, requestsInOrder.Count);
+            
+            var index = 0;
+            var cancelled = 0;
+            var dequeueTasks = new ConcurrentBag<Task<RequestMessage>>();
+
+            var cancelSomeTask = Task.Run(() =>
+            {
+                foreach (var _ in requestsInOrder)
+                {
+                    var currentIndex = Interlocked.Increment(ref index);
+
+                    if(currentIndex % 2 == 0)
+                    {
+                        Interlocked.Increment(ref cancelled);
+                        queueAndWaitTasksInOrder.ElementAt(index-1).Item2.Cancel();
+                    }
+                }
+            });
+
+            // Cancellation is slow so give it a head start
+            if (cancelled < 100)
```

Comment:
Should this be a loop?

Reply:
Haha, it should!! i wondered why it was a bit crap at cancelling 😆